In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df=pd.read_csv("data.csv",encoding="ISO-8859-1")

In [72]:
df.tail(1)

Date  Label                                               Top1  \
4100  2016-07-01      1  A 117-year-old woman in Mexico City finally re...   

                                                  Top2  \
4100  IMF chief backs Athens as permanent Olympic host   

                                                   Top3  \
4100  The president of France says if Brexit won, so...   

                                                   Top4  \
4100  British Man Who Must Give Police 24 Hours' Not...   

                                                   Top5  \
4100  100+ Nobel laureates urge Greenpeace to stop o...   

                                                   Top6  \
4100  Brazil: Huge spike in number of police killing...   

                                                   Top7  \
4100  Austria's highest court annuls presidential el...   

                                                   Top8  ...  \
4100  Facebook wins privacy case, can track any Belg...  ...   

                                                  Top16  \
4100  The United States has placed Myanmar, Uzbekist...   

                                                  Top17  \
4100  S&amp;P revises European Union credit rating t...   

                                                  Top18  \
4100  India gets $1 billion loan from World Bank for...   

                                                  Top19  \
4100  U.S. sailors detained by Iran spoke too much u...   

                                                  Top20  \
4100  Mass fish kill in Vietnam solved as Taiwan ste...   

                                                  Top21  \
4100  Philippines president Rodrigo Duterte urges pe...   

                                                  Top22  \
4100  Spain arrests three Pakistanis accused of prom...   

                                                  Top23  \
4100  Venezuela, where anger over food shortages is ...   

                                                  Top24  \
4100  A Hindu temple worker has been killed by three...   

                                                  Top25  
4100  Ozone layer hole seems to be healing - US &amp...  

[1 rows x 27 columns]

In [50]:
train=df[:][:3500]
test=df[:][3500:]

In [52]:
#removing anything except [a-zA-Z],i.e any punctuator with " "
data=train.drop(["Date","Label"],axis=1)
data.replace('[^a-zA-Z]'," ",regex=True,inplace=True)

In [53]:
#renaming all columns with numerical value
name=[i for i in range(1,26)]
data.columns=name
data.head(1)

1          2   \
0  A  hindrance to operations   extracts from the...  Scorecard   

                                3                                         4   \
0  Hughes  instant hit buoys Blues  Jack gets his skates on at ice cold Alex   

                                       5   \
0  Chaos as Maracana builds up for United   

                                                  6   \
0  Depleted Leicester prevail as Elliott spoils E...   

                                 7                                  8   \
0  Hungry Spurs sense rich pickings  Gunners so wide of an easy target   

                                              9   \
0  Derby raise a glass to Strupar s debut double   

                                         10  ...  \
0  Southgate strikes  Leeds pay the penalty  ...   

                                         16  \
0  Flintoff injury piles on woe for England   

                                                  17  \
0  Hunters threaten Jospin with new battle of the...   

                                    18                                    19  \
0  Kohl s successor drawn into scandal  The difference between men and women   

                                    20  \
0  Sara Denver  nurse turned solicitor   

                                               21  \
0  Diana s landmine crusade put Tories in a panic   

                                                  22                23  \
0  Yeltsin s resignation caught opposition flat f...  Russian roulette   

         24                  25  
0  Sold out  Recovering a title  

[1 rows x 25 columns]

In [54]:
#for save text pre processing...here we converted all text in dataset as small case text
for feature in data.columns:
    data[feature]=data[feature].str.lower()      
data.head(1)

1          2   \
0  a  hindrance to operations   extracts from the...  scorecard   

                                3                                         4   \
0  hughes  instant hit buoys blues  jack gets his skates on at ice cold alex   

                                       5   \
0  chaos as maracana builds up for united   

                                                  6   \
0  depleted leicester prevail as elliott spoils e...   

                                 7                                  8   \
0  hungry spurs sense rich pickings  gunners so wide of an easy target   

                                              9   \
0  derby raise a glass to strupar s debut double   

                                         10  ...  \
0  southgate strikes  leeds pay the penalty  ...   

                                         16  \
0  flintoff injury piles on woe for england   

                                                  17  \
0  hunters threaten jospin with new battle of the...   

                                    18                                    19  \
0  kohl s successor drawn into scandal  the difference between men and women   

                                    20  \
0  sara denver  nurse turned solicitor   

                                               21  \
0  diana s landmine crusade put tories in a panic   

                                                  22                23  \
0  yeltsin s resignation caught opposition flat f...  russian roulette   

         24                  25  
0  sold out  recovering a title  

[1 rows x 25 columns]

In [55]:
#Now we'll try combining all the 25 column headline(Text) of a single row...and do the same for all rows
combined=[]     ##will be furthur used for storing combined headlines
# Combined=[headline+headline  for row in range(0,len(data.index)) for headline in data.iloc[row,0:25]]
headline=""
for row in range(0,len(data.index)):
    for txt in data.iloc[row,0:25]:
        headline=headline+str(txt)+" "
    combined.append(headline)
    headline=""
    
combined[0]

'a  hindrance to operations   extracts from the leaked reports scorecard hughes  instant hit buoys blues jack gets his skates on at ice cold alex chaos as maracana builds up for united depleted leicester prevail as elliott spoils everton s party hungry spurs sense rich pickings gunners so wide of an easy target derby raise a glass to strupar s debut double southgate strikes  leeds pay the penalty hammers hand robson a youthful lesson saints party like it s      wear wolves have turned into lambs stump mike catches testy gough s taunt langer escapes to hit     flintoff injury piles on woe for england hunters threaten jospin with new battle of the somme kohl s successor drawn into scandal the difference between men and women sara denver  nurse turned solicitor diana s landmine crusade put tories in a panic yeltsin s resignation caught opposition flat footed russian roulette sold out recovering a title '

# USING COUNT VECTORIZER
for all of those who don't know...since machines do not know the text language and here we are dealing with plain gramatical words...so using NPL(Natural Language Processing) we'll be assigning a new column to each and every unique word and when found in an element we'll just be assigning 1 for it..else 0.
And thus in this way we'll be overcomming this plain text problem
And the module which help's us in doing so is COUNT VECTORISER

In [56]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.ensemble import RandomForestClassifier

here what ngram_range means is that...after all unique words forming a new column...now 2x2 permutation of words will be formed and added as new columns

In [57]:
#fitting
cv=CountVectorizer(ngram_range=(2,2))
train_data_cv=cv.fit_transform(combined)

In [60]:
rc=RandomForestClassifier(n_estimators=200,criterion='entropy')
rc.fit(train_data_cv,train.Label)

RandomForestClassifier(criterion='entropy', n_estimators=200)

# NLP with Test data
Now working on test data...wrangling it too

In [61]:
#droping unnecessary columns in test data and removing any non-alphabetical character 
data=test.drop(['Date','Label'],axis=1)
data.replace("[^a-zA-Z]"," ",regex=True,inplace=True)

In [62]:
#renaming all columns with numerical value
name=[i for i in range(1,26)]
data.columns=name
data.head(1)

1   \
3500  London City Banker tried to put neighbour in m...   

                                                     2   \
3500  Italy court overturns law equating cannabis wi...   

                                                3   \
3500  Venezuela anti government rally turns deadly   

                                                     4   \
3500  Slaughtered for Ivory      of Forest Elephants...   

                                                     5   \
3500  Massive Protests in Venezuela ignored by local...   

                                                     6   \
3500  The risk of reporting US drone strikes  Yemen ...   

                                                     7   \
3500  Mexican Tycoon Becomes First Foreigner to Expl...   

                                                     8   \
3500  Britain warns Scotland  Forget the pound if yo...   

                                                     9   \
3500  UK launches         fund to help male victims ...   

                                                     10  ...  \
3500  German postwoman sacked for taking too much ti...  ...   

                                                     16  \
3500  EU tells Britain to include Prostitutes And Dr...   

                                                     17  \
3500  Venezuela  One Wounded  Two Dead after Opposit...   

                                                     18  \
3500  Chinese villagers attack factory after reports...   

                                                     19  \
3500  Strange  S  shaped phenomenon appears off West...   

                                            20  \
3500  China announces     B air pollution fund   

                                                     21  \
3500  The Bank of England has confirmed that economi...   

                                            22  \
3500  Male prostitution on the rise in Lebanon   

                                                     23  \
3500  Indian Parliament erupts in chaos after MP use...   

                                           24  \
3500  Spain makes abortion a criminal offense   

                                                     25  
3500  Nuclear fusion breakthrough raises hopes for u...  

[1 rows x 25 columns]

In [63]:
#for save text pre processing...here we converted all text in dataset as small case text
for feature in data.columns:
    data[feature]=data[feature].str.lower()    

In [64]:
# Now we'll try combining all the 25 column headline(Text) of a single row...and do the same for all rows
combined=[]     ##will be furthur used for storing combined headlines
# Combined=[headline+headline  for row in range(0,len(data.index)) for headline in data.iloc[row,0:25]]
headline=""
for row in range(0,len(data.index)):
    for txt in data.iloc[row,0:25]:
        headline=headline+str(txt)+" "
    combined.append(headline)
    headline=""
    


# Applying count_vectorizer in test data

In [65]:
cv_test=CountVectorizer(vocabulary=cv.get_feature_names(),ngram_range=(2,2))

test_data_cv=cv_test.transform(combined)

C:\Python37\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


In [66]:
test_data_cv.shape

(601, 501554)

In [67]:
predictions=rc.predict(test_data_cv)

In [68]:
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report

In [69]:
print(confusion_matrix(test.Label,predictions))

[[ 29 253]
 [ 35 284]]


In [70]:
print(accuracy_score(test.Label,predictions))

0.5207986688851913


In [71]:
print(classification_report(test.Label,predictions))

              precision    recall  f1-score   support

           0       0.45      0.10      0.17       282
           1       0.53      0.89      0.66       319

    accuracy                           0.52       601
   macro avg       0.49      0.50      0.42       601
weighted avg       0.49      0.52      0.43       601



Now many of you might contradict my point of approach which i too think is not yet the best one...but a good one to start with if working on NLP.
Here i should have vectorised whole of dataset after cleaning and wrangling and then sliced it for train andd test data.And this i blv would have been the best approach.
But still we achieved an accuracy of 52% which i wouldn't say is the best...but good one for an NLP project.